File: ideaproject/javaserializer/src/com/houseelectrics/serializer/Object2Json.java
package com.houseelectrics.serializer;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;

import java.util.Stack;

/**
 * Created by roberttodd on 30/11/2014.
 */
public class Object2Json implements ExplorationListener
{

    public interface ExplorerFactory
    {
        public Explorer create();
    }

    public Object2Json()
    {
        explorerFactory= new ExplorerFactory()
        {
            @Override
            public Explorer create()
            {
                return new ObjectExplorerImpl();
            }
        };
    }

    private boolean alwaysDoubleQuotePropertyNames = false;
    public boolean getAlwaysDoubleQuotePropertyNames() {return alwaysDoubleQuotePropertyNames;}
    public void setAlwaysDoubleQuotePropertyNames(boolean value) {this.alwaysDoubleQuotePropertyNames=value;}

    private TypeAliaser typeAliaser=null;
    public TypeAliaser getTypeAliaser() {return typeAliaser;}
    public void setTypeAliaser(TypeAliaser value) {this.typeAliaser=value;}

    private boolean omitMarkAsArrayFunction = true;
    public boolean getOmitMarkAsArrayFunction() {return omitMarkAsArrayFunction;}
    public void setOmitMarkAsArrayFunction(boolean value) {this.omitMarkAsArrayFunction=value;}

    private boolean omitDefaultLeafValuesInJs = false;
    public boolean getOmitDefaultLeafValuesInJs() {  return omitDefaultLeafValuesInJs; }
    public void setOmitDefaultLeafValuesInJs(boolean value) { this.omitDefaultLeafValuesInJs = value; }

    private LeafDefaultSet leafDefaultSet = null;
    public LeafDefaultSet getLeafDefaultSet() {  return this.leafDefaultSet; }
    public void setLeafDefaultSet(LeafDefaultSet value) { this.leafDefaultSet = value; }

    String objectResolverFunctionName = "resolveRefs";
    public String getObjectResolverFunctionName() { return objectResolverFunctionName; }
    public void setObjectResolverFunctionName(String value) { this.objectResolverFunctionName = value; }

    String leafDefaultResolverFunctionName = "_resolveLeafDefaults";
    public String getLeafDefaultResolverFunctionName() { return leafDefaultResolverFunctionName; }
    public void setLeafDefaultResolverFunctionName(String value) { this.leafDefaultResolverFunctionName = value; }

    String attachId2ArrayFunctionName="attachId2Array";
    public String getAttachId2ArrayFunctionName() { return attachId2ArrayFunctionName; }
    public void setAttachId2ArrayFunctionName(String value) { this.attachId2ArrayFunctionName = value; }
    public String getAttachId2ArrayJSFunction()
    {
        return "function " + getAttachId2ArrayFunctionName() +  "(id, arr) {arr." + idTag + "=id; arr." + isArrayIndicator + "='y'; return arr;}";
    }

    String markAsArrayFunctionName = "_a_";
    public String getMarkAsArrayFunctionName() { return markAsArrayFunctionName; }
    public void setMarkAsArrayFunctionName(String value) { markAsArrayFunctionName = value; }
    public String getMarkAsArrayJSFunction()
    {
        return Object2JsonJavascript.getMarkAsArrayJSFunction(getMarkAsArrayFunctionName());
    }

    String idTag = "_id_";
    String getIdTag() { return idTag; }
    public void setIdTag(String value) { idTag = value; }
    String referenceTag = "_ref_";
    String getReferenceTag()  { return referenceTag; }
    public void setReferenceTag(String value) { referenceTag = value; }
    boolean useReferences = false;
    public boolean getUseReferences()  { return useReferences; }
    public void setUseReferences(boolean value) { useReferences = value; }
    public final String NewLine = "\r\n";
    NodeExpander nodeExpander;
    public NodeExpander getNodeExpander() {return nodeExpander;}
    public void setNodeExpander(NodeExpander value) {this.nodeExpander=value;}
    public String toJson(Object o)
    {
        return toJson(o, getTypeAliaser());
    }

    public String toJson(Object o, TypeAliaser typeAliaser)
    {
       try
       {
           StringWriter sw = new StringWriter();
           writeAsJson(o, sw, typeAliaser);
           sw.flush();
           String str = sw.toString();
           sw.close();
           return str;
       }
       catch (IOException ioex)
       {
           throw new RuntimeException("io failed to write json", ioex);
       }
    }

    ExplorerFactory explorerFactory;
    public void setExplorerFactory(ExplorerFactory value) { explorerFactory = value; }
    public ExplorerFactory getExplorerFactory() { return explorerFactory; }

    private void writePropertyName(String propertyName, Writer writer) throws IOException
    {
        boolean escapeRequired = this.getAlwaysDoubleQuotePropertyNames() || (propertyName.indexOf('.') >= 0);
        if (escapeRequired) writer.write('\"');
        writer.write(propertyName);
        if (escapeRequired) writer.write('\"');
    }

    public boolean MoveAway(Object from, String propertyName, Object to, boolean isIndexed, Integer index)
    {
        try
        {
            return MoveAwayUnsafe(from, propertyName, to, isIndexed, index);
        }
        catch (Exception ex)
        {
            throw new RuntimeException("Mode away failed", ex);
        }
    }

    public  boolean MoveAwayUnsafe(Object from, String propertyName, Object to, boolean isIndexed, Integer index) throws IOException
    {
        ExploreStackFrame currentFrame = exploreStack.size()>0 ? exploreStack.peek():null;
        if (currentFrame != null)
        {
            if (currentFrame.propertyCount > 0) writer.write(", ");
            currentFrame.propertyCount++;
        }
        if (from != null && propertyName != null)
        {
            writePropertyName(propertyName, writer);
            writer.write(":");
        }
        ExploreStackFrame childFrame = new ExploreStackFrame();
        exploreStack.push(childFrame);
        writeIndent(writer, exploreStack);
        if (getUseReferences())
        {

            ObjectIDGenerator.IdResult idResult = idGenerator.getId(to);
            int objectid = idResult.id;
            boolean firstTime = idResult.isNew;
            if (firstTime)
            {
                // could be done like this ! (function() {var x=[1,2]; x.id="uuu";return x;})()
                if (!isIndexed)
                {
                    writer.write("{" + this.getIdTag() + ":" + objectid + ' ');
                    childFrame.propertyCount++;
                    childFrame.propertyCount += writeTypeAliasProperty(writer, to, typeAliaser, childFrame.propertyCount);

                }
                else
                {
                    // no need for type alias
                    writer.write(getAttachId2ArrayFunctionName() + "(" + objectid + ",[");
                }

            }
            else
            {
                writer.write("{" + this.getReferenceTag() + ":" + objectid);
                return false;
            }
        }
        else // !Use References
        {
            if (!isIndexed)
            {
                writer.write('{');
                // todo -- check this out ............
                childFrame.propertyCount += writeTypeAliasProperty(writer, to, typeAliaser, childFrame.propertyCount);
            }
            else
            {
                if (!getOmitMarkAsArrayFunction())
                {
                    writer.write(markAsArrayFunctionName);
                    writer.write("([");
                }
                else
                {
                    writer.write("[");
                }
            }
        }

        return true;
    }

    public  void MoveBack(Object from, String propertyName, Object to, boolean isIndexed)
    {
        try
        {
            if (!isIndexed) writer.write('}');
            else
            {
                writer.write(']');
                // is there a function wrapper ?
                if (!getOmitMarkAsArrayFunction() || getUseReferences()) writer.write(")");
            }
            exploreStack.pop();
            writeIndent(writer, exploreStack);
        }
        catch (IOException ioex)
        {
            throw new RuntimeException("failed to move back from " + from.getClass().getName() + "." + propertyName);
        }
    }

    public  void OnLeaf(Object from, String propertyName, Object to, Integer index)
    {
        try {
            OnLeafUnsafe(from,  propertyName,  to,  index);
        }
        catch (IOException ioex)
        {
            throw new RuntimeException("OnLeaf failed for " + from.getClass().getName() + "." + propertyName + "->" + to + " index=" + index);
        }
    }

    public  void OnLeafUnsafe(Object from, String propertyName, Object to, Integer index) throws IOException
    {
        //check for default leaf values
        if (!this.getOmitDefaultLeafValuesInJs() ||
                !isDefaultLeafValue(from, propertyName, to, getLeafDefaultSet()))
        {
            ExploreStackFrame currentFrame = exploreStack.peek();
            if (currentFrame.propertyCount > 0) writer.write(", ");
            currentFrame.propertyCount++;
            if (propertyName!=null)
            {
                writePropertyName(propertyName, writer);
                writer.write(":");
            }
            leafWriter.writeLeafValue(writer, to, propertyName);
        }

    }

    ObjectIDGenerator idGenerator = null;

class ExploreStackFrame { public int propertyCount;}

    public void writeAsJson(Object o, StringWriter writer)
    {
        writeAsJson(o, writer, null);
    }


    Stack<ExploreStackFrame> exploreStack = null;

    Writer writer;

    public void writeAsJson(Object o, StringWriter writer, TypeAliaser typeAliaser)
    {
        if (getLeafDefaultSet() != null && getOmitDefaultLeafValuesInJs() && getOmitMarkAsArrayFunction())
        {
            throw new RuntimeException("Leaf defaulting requires Array marker for js code");
        }
        this.writer = writer;
        //this.typeAliaser = typeAliaser;

         idGenerator = null;

        if (getUseReferences())
        {
            idGenerator=new ObjectIDGenerator();
        }
        exploreStack = new Stack<ExploreStackFrame>();
        Explorer explorerImpl = getExplorerFactory().create();
        ((Explorer)explorerImpl).setNodeExpander(getNodeExpander());

        explorerImpl.explore(o, this);
    }

    int  writeTypeAliasProperty(Writer writer, Object to, TypeAliaser typeAliaser, int currentFramePropertyCount) throws IOException
    {
        if (typeAliaser != null)
        {
            if (currentFramePropertyCount > 0) writer.write(',');
            writer.write(this.getTypeAliasProperty() + ":'" + typeAliaser.alias (to.getClass()) + "'");
            return 1;
        }
        else
        {
            return 0;
        }
    }

    //public Func<Object, String, Object, LeafDefaultSet, boolean> isDefaultLeafValue = isDefaultLeafValueDefault;

    protected boolean isDefaultLeafValue(Object from, String propertyName, Object leafValue, LeafDefaultSet LeafDefaultSet)
    {
        if ( LeafDefaultSet == null)
        {
            throw new RuntimeException("OmitDefaultLeafValuesInJs==true but LeafDefaultSet is unspecified - unable to determine what values are default !");
        }

        Class t = from.getClass();
        if (LeafDefaultSet.getType2Defaults().containsKey(t))
        {
            Defaults4Class defaults = LeafDefaultSet.getType2Defaults().get(t);
            if (!defaults.getPropertyName2DefaultValue().containsKey(propertyName))
            {
                return false;
            }
            Object defaultValue = defaults.getPropertyName2DefaultValue().get(propertyName);
            return (leafValue == defaultValue) || leafValue != null && leafValue.equals(defaultValue);
        }
        {
            // is this an error ?
            return false;
        }
    }

    void writeIndent(Writer writer, Stack<ExploreStackFrame> exploreStack) throws IOException
    {
        if (getIndentSize() != null)
        {
            writer.write(NewLine);
            int fullIndentSize = (getIndentSize().intValue()) * exploreStack.size();
            for (int done = 0; done < fullIndentSize; done++)
            {
                writer.write(' ');
            }
        }
    }

    private String isArrayIndicator = "_a_";

    public interface LeafWriter
    {
        public void writeLeafValue(Writer writer, Object to, String propertyName) throws IOException;
    }

    private LeafWriter defaultLeafWriter = new LeafWriter()
    {
        @Override
        public void writeLeafValue(Writer writer, Object to, String propertyName) throws IOException
        {
            if (to == null)
            {
                writer.write("null");
            }
            else if (to.getClass() == String.class || to.getClass() == Character.class)
            {
                writer.write("\"");
                writer.write(to.toString());
                writer.write("\"");
            }//todo simplify below
            else if (to.getClass() == Byte.class)
            {
                writer.write(to.toString());
            }
            else if (to.getClass() == Integer.class)
            {
                writer.write(to.toString());
            }
            else if (to.getClass() == Long.class)
            {
                writer.write(((Long)to).toString());
            }
            else if (to.getClass() == Float.class)
            {
                writer.write(((Float)to).toString());
            }
            else if (to.getClass() == Double.class)
            {
                writer.write(((Double)to).toString());
            }

            else if (to.getClass() == Boolean.class)
            {
                boolean bval = ((Boolean) to).booleanValue();
                writer.write(bval ? "true" : "false");
            }
            else if (to.getClass() == Short.class)
            {
                writer.write(((Short)to).toString());
            }
            else throw new RuntimeException("not implemented writeLeafValue:" + to.getClass().getName() + " " + propertyName);

        }
    };

    public LeafWriter getDefaultLeafWriter() {return defaultLeafWriter;}
    private LeafWriter leafWriter = getDefaultLeafWriter();
    public void setLeafWriter(LeafWriter value)
    {
        this.leafWriter = value;
    }
    public LeafWriter getLeafWriter()
    {
        return this.leafWriter;
    }


    private Integer indentSize=null;
    public Integer getIndentSize() { return indentSize; }
    public void setIndentSize(Integer value) { indentSize = value; }

    public String  getObjectResolverJS()
    {
        return Object2JsonJavascript.getObjectResolverJS(getObjectResolverFunctionName(), getIdTag(), getReferenceTag());
    }

    String typeAliasProperty = "_t_";
    public String getTypeAliasProperty() {return typeAliasProperty; }
    public void setTypeAliasProperty(String value) { this.typeAliasProperty = value; }

    public String getLeafDefaultResolverJS()
    {
        return getLeafDefaultResolverJS(getTypeAliaser());
    }

    public String getLeafDefaultResolverJS(TypeAliaser typeAliaser)
    {
        //switch off type aliasing

        String defaultsJS = toJson( getLeafDefaultSet().getAlias2Defaults(typeAliaser), null);
        String defaultsVarName = "_typeAlias2LeafDefaults_";

        String js =
        "var "+ defaultsVarName +" = " + defaultsJS + ";\r\n" +
         Object2JsonJavascript.ResolveLeafDefaultFunction;
        return js;
    }

}

File: ideaproject/javaserializer/src/com/houseelectrics/serializer/Object2Json.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using StreamWriter = System.IO.StreamWriter;
using TextWriter = System.IO.TextWriter;
using StringWriter = System.IO.StringWriter;
using ObjectIDGenerator = System.Runtime.Serialization.ObjectIDGenerator;
//search for ^ reg ex- gives 1776 29 july 2013 of which 1026 test
//search for ^ reg ex- gives 1082 14 june 2013 of which 657 test
//search for ^ reg ex- gives 949 16 june 2013 of which 521 test, 428 serializer
//TODO add unit test for AlwaysDoubleQuotePropertyNames

namespace com.houseelectrics.serializer 
{
    public class Object2Json
    {
        private bool alwaysDoubleQuotePropertyNames = false;
        public bool AlwaysDoubleQuotePropertyNames
           {
              get { return alwaysDoubleQuotePropertyNames; }
              set { this.alwaysDoubleQuotePropertyNames = value; }
           }

        private TypeAliaser typeAliaser=null;
        public TypeAliaser TypeAliaser { get { return typeAliaser; } set { typeAliaser = value; } }

        private bool omitMarkAsArrayFunction = true;
        public bool OmitMarkAsArrayFunction { get { return omitMarkAsArrayFunction; } set { omitMarkAsArrayFunction = value; } }

        
        private bool omitDefaultLeafValuesInJs = false;
        public bool OmitDefaultLeafValuesInJs { get { return omitDefaultLeafValuesInJs; } set { omitDefaultLeafValuesInJs = value; } }

        private LeafDefaultSet leafDefaultSet = null;
        public LeafDefaultSet LeafDefaultSet { get { return leafDefaultSet; } set { leafDefaultSet = value; } }


        string objectResolverFunctionName = "resolveRefs";
        public string ObjectResolverFunctionName { get { return objectResolverFunctionName; } set { this.objectResolverFunctionName = value; } }

        string leafDefaultResolverFunctionName = "_resolveLeafDefaults";
        public string LeafDefaultResolverFunctionName { get { return leafDefaultResolverFunctionName; } set { this.leafDefaultResolverFunctionName = value; } }

        string attachId2ArrayFunctionName="attachId2Array";
        public string AttachId2ArrayFunctionName { get { return attachId2ArrayFunctionName; } set { attachId2ArrayFunctionName = value; } }
        public String getAttachId2ArrayJSFunction()
        {
            return "function " + AttachId2ArrayFunctionName +  "(id, arr) {arr." + idTag + "=id; arr." + isArrayIndicator + "='y'; return arr;}";
        }

        string markAsArrayFunctionName = "_a_";
        public string MarkAsArrayFunctionName { get { return markAsArrayFunctionName; } set { markAsArrayFunctionName = value; } }
        public String getMarkAsArrayJSFunction()
        {
            return Object2JsonJavascript.getMarkAsArrayJSFunction(MarkAsArrayFunctionName); 
        }   

        string idTag = "_id_";
        string IdTag { get { return idTag; } set { idTag = value; } }
        string referenceTag = "_ref_";
        string ReferenceTag { get { return referenceTag; } set { referenceTag = value; } }
        bool useReferences = false;
        public bool UseReferences { get { return useReferences; } set { useReferences = value; } }
        const string NewLine = "\r\n";
        public NodeExpander NodeExpander {get; set;}
        public string toJson(Object o)
        {
            return toJson(o, TypeAliaser);
        }

        public string toJson(Object o, TypeAliaser typeAliaser)
        {
            /**
                o2J.OmitDefaultLeafValuesInJs = true;
                o2J.LeafDefaultSet = lds;
            //  to do throw exception if this is not the case
            //  o2J.OmitMarkAsArrayFunction = false;
           **/

         
            StringWriter sw = new StringWriter();
            writeAsJson(o, sw, typeAliaser);
            sw.Flush();
            String str = sw.ToString();
            sw.Close();
            return str;
        }

        Func<Explorer> explorerFactory = () => { return new ObjectExplorerImpl(); };

        public Func<Explorer> ExplorerFactory { get { return explorerFactory; } set { explorerFactory = value; } }

        private void writePropertyName(string propertyName, TextWriter writer)
        {
            bool escapeRequired = this.AlwaysDoubleQuotePropertyNames || (propertyName.IndexOf('.') >= 0);
            if (escapeRequired) writer.Write('\"');
            writer.Write(propertyName);
            if (escapeRequired) writer.Write('\"');
        }

        class ExploreStackFrame { public int propertyCount;}
        public void writeAsJson(Object o, TextWriter writer, TypeAliaser typeAliaser=null)
        {
            if (LeafDefaultSet != null && OmitDefaultLeafValuesInJs && OmitMarkAsArrayFunction)
            {
                throw new Exception("Leaf defaulting requires Array marker for js code");
            }


            ObjectIDGenerator idGenerator = null;

            if (UseReferences)
            {
                idGenerator=new ObjectIDGenerator();
            }
            Stack<ExploreStackFrame> exploreStack = new Stack<ExploreStackFrame>();
            Explorer explorerImpl = ExplorerFactory();
            ((Explorer)explorerImpl).NodeExpander = NodeExpander;
            MoveAway down = delegate (Object from, string propertyName, Object to, bool isIndexed, int? index)
            {
                ExploreStackFrame currentFrame = exploreStack.Count>0 ? exploreStack.Peek():null;
                if (currentFrame != null)
                {
                    if (currentFrame.propertyCount > 0) writer.Write(", ");
                    currentFrame.propertyCount++;
                }
                if (from != null && propertyName != null)
                {
                    writePropertyName(propertyName, writer);
                    writer.Write(":");
                }
                ExploreStackFrame childFrame = new ExploreStackFrame();
                exploreStack.Push(childFrame);
                writeIndent(writer, exploreStack);
                if (UseReferences)
                {
                    bool firstTime;
                    long objectid = idGenerator.GetId(to, out firstTime);
                    if (firstTime)
                    {
                        // could be done like this ! (function() {var x=[1,2]; x.id="uuu";return x;})()
                        if (!isIndexed)
                        {
                            writer.Write("{" + this.IdTag + ":" + objectid + ' ');
                            childFrame.propertyCount++;
                            childFrame.propertyCount += writeTypeAliasProperty(writer, to, typeAliaser, childFrame.propertyCount);
                            
                        }
                        else
                        {
                            // no need for type alias
                            writer.Write(AttachId2ArrayFunctionName + "(" + objectid + ",[");
                        }
                        
                    }
                    else
                    {
                        writer.Write("{" + this.ReferenceTag + ":" + objectid);
                        return false;
                    }
                }
                else // !Use References
                {
                    if (!isIndexed)
                    {
                        writer.Write('{');
                        // todo -- check this out ............
                        childFrame.propertyCount += writeTypeAliasProperty(writer, to, typeAliaser, childFrame.propertyCount);
                    }
                    else
                    {
                        if (!OmitMarkAsArrayFunction)
                        {
                            writer.Write(markAsArrayFunctionName);
                            writer.Write("([");
                        }
                        else
                        {
                            writer.Write("[");
                        }
                    }
                }

                return true;
            };

            MoveBack up = (from, propertyName, to, isIndexed) => 
                {
                    if (!isIndexed) writer.Write('}');
                    else
                    {
                        writer.Write(']');
                        // is there a function wrapper ?
                        if (!OmitMarkAsArrayFunction || UseReferences) writer.Write(")");
                    }
                    exploreStack.Pop();
                    writeIndent(writer, exploreStack);
                };

            OnLeaf leaf = (from, propertyName, to, index) => 
              {
                  //check for default leaf values
                  if (!this.OmitDefaultLeafValuesInJs ||
                      !isDefaultLeafValue(from, propertyName, to, LeafDefaultSet))
                  {
                     ExploreStackFrame currentFrame = exploreStack.Peek();
                     if (currentFrame.propertyCount > 0) writer.Write(", ");
                     currentFrame.propertyCount++;
                     if (propertyName!=null)
                     {
                         writePropertyName(propertyName, writer);
                         writer.Write(":");                     
                     }
                  writeLeafValue(writer, to, propertyName);
                  }
              };
            explorerImpl.explore(o, down, up, leaf);
        }

        int  writeTypeAliasProperty(TextWriter writer, Object to, TypeAliaser typeAliaser, int currentFramePropertyCount)
        {
            if (typeAliaser != null)
            {
                if (currentFramePropertyCount > 0) writer.Write(',');
                writer.Write(this.TypeAliasProperty + ":'" + typeAliaser(to.GetType()) + "'");
                return 1;
            }
            else
            {
                return 0;
            }
        }

        public Func<object, string, object, LeafDefaultSet, bool> isDefaultLeafValue = isDefaultLeafValueDefault;

        static bool isDefaultLeafValueDefault(object from, string propertyName, object leafValue, LeafDefaultSet LeafDefaultSet)
        {
            if ( LeafDefaultSet == null)
            {
                throw new Exception("OmitDefaultLeafValuesInJs==true but LeafDefaultSet is unspecified - unable to determine what values are default !");
            }

            Type t = from.GetType();
            if (LeafDefaultSet.Type2Defaults.ContainsKey(t))
            {
                Defaults4Class defaults = LeafDefaultSet.Type2Defaults[t];
                if (!defaults.PropertyName2DefaultValue.ContainsKey(propertyName))
                {
                    return false;
                }
                Object defaultValue = defaults.PropertyName2DefaultValue[propertyName];
                return (leafValue == defaultValue) || leafValue != null && leafValue.Equals(defaultValue);
            }
            {
                // is this an error ?
                return false;
            }
        }

        void writeIndent(TextWriter writer, Stack<ExploreStackFrame> exploreStack)
        {
            if (IndentSize != null)
            {
                writer.Write(NewLine);
                int fullIndentSize = ((int)IndentSize) * exploreStack.Count;
                for (int done = 0; done < fullIndentSize; done++)
                {
                    writer.Write(' ');
                }
            }
        }

        private string isArrayIndicator = "_a_";

        public Action<TextWriter, Object, string> writeLeafValue = defaultWriteLeafValue;

        public static Action<TextWriter, Object, string> defaultWriteLeafValue = (writer, to, propertyName) =>
        {
            if (to == null)
            {
                writer.Write("null");
            }
            else if (to.GetType() == typeof(string) || to.GetType() == typeof(Char))
            {
                writer.Write("\"");
                writer.Write(to.ToString());
                writer.Write("\"");
            }//todo simplify below
            else if (to.GetType() == typeof(Byte))
            {
                writer.Write((Byte)to);
            }
            else if (to.GetType() == typeof(Int16))
            {
                writer.Write((Int16)to);
            }
            else if (to.GetType() == typeof(Int32))
            {
                writer.Write((Int32)to);
            }
            else if (to.GetType() == typeof(Int64))
            {
                writer.Write((Int64)to);
            }
            else if (to.GetType() == typeof(UInt16))
            {
                writer.Write((UInt16)to);
            }
            else if (to.GetType() == typeof(UInt32))
            {
                writer.Write((UInt32)to);
            }
            else if (to.GetType() == typeof(UInt64))
            {
                writer.Write((UInt64)to);
            }
            else if (to.GetType() == typeof(Double))
            {
                writer.Write((Double)to);
            }
            else if (to.GetType() == typeof(Single))
            {
                writer.Write((Single)to);
            }
            else if (to.GetType() == typeof(Boolean))
            {
                writer.Write(((Boolean)to).CompareTo(true) == 0 ? "true" : "false");
            }

            else throw new Exception("not implemented writeLeafValue:" + to.GetType() + " " + propertyName);


        };

        private int? indentSize=null;
        public int? IndentSize { get { return indentSize; } set { indentSize = value; } }

    public string  getObjectResolverJS()
    {
        return Object2JsonJavascript.getObjectResolverJS(ObjectResolverFunctionName, IdTag, ReferenceTag);
    }

    string typeAliasProperty = "_t_";
    public string TypeAliasProperty { get { return typeAliasProperty; } set { this.typeAliasProperty = value; } }

    public string getLeafDefaultResolverJS()
    {
        return getLeafDefaultResolverJS(TypeAliaser);
    }

    public string getLeafDefaultResolverJS(TypeAliaser typeAliaser)
    {
        //switch off type aliasing
        
        string defaultsJS = toJson( LeafDefaultSet.getAlias2Defaults(typeAliaser), null);
        string defaultsVarName = "_typeAlias2LeafDefaults_";

        string js =
@"var "+ defaultsVarName +" = " + defaultsJS + @";
" + Object2JsonJavascript.ResolveLeafDefaultFunction;           
        return js;
    }
    }
}

/*
 * DONE
 improve browser test page 
 refactor js functions
 refactor test js functions
 done implement defaulting
 done fix array detection in Object resolver js
 done fit test results onto 1 test / example page
 done deserializer
 done implement security
 done asp examples
** TODO
   doneuserguide      
   replace use of ms script engine for tests Microsoft.JScript.Vsa.VsaEngine, Eval.JScriptEvaluate(strJS, vsaEngine);
   test on different browsers
       : testing todds.me.uk, ftp user u48650861, http://browsershots.org/
   review parameterization of js
   integration with NSpring
   maven build
   open source submission
   redo for java - more popular ?
 */
